File: source/resources/tail-recursive-optimization/tailrec.cpp
#include <iostream>

using namespace std;

int sum(int n) {
    if (n <= 1) return n;
    return n + sum(n-1);
}

int tailsum(int n, int sum) {
    if (n == 0) return sum;
    return tailsum(n-1, sum+n);
}

int main() {

    std::cout << sum(100) << std::endl;
    std::cout << sum(1000000) << std::endl;
    //std::cout << tailsum(1000000, 0) << std::endl;
    return 0;
}

File: resources/tail-recursive-optimization/TailRec.java
public class TailRec {


    public static int tailSum(int n, int sum) {
        if (n == 0) return sum;
        return tailSum(n-1, n+sum);
    }

    public static void main(String[] args) {
        System.out.println(tailSum(100, 0));
        System.out.println(tailSum(100000, 0));
    }
}
